rclcpp/include/rclcpp/executor.hpp

```diff
@@ -319,6 +320,50 @@ class Executor
   virtual void
   spin_once(std::chrono::nanoseconds timeout = std::chrono::nanoseconds(-1));
 
+  /// Spin (blocking) until the condition is complete, it times out waiting,
+  /// or rclcpp is interrupted.
+  /**
+   * \param[in] future The condition which can be callable or future type to wait on.
+   *   If this function returns SUCCESS, the future can be
+   *   accessed without blocking (though it may still throw an exception).
+   * \param[in] timeout Optional timeout parameter, which gets passed to Executor::spin_node_once.
+   *   `-1` is block forever, `0` is non-blocking.
+   *   If the time spent inside the blocking loop exceeds this timeout, return a TIMEOUT return
+   *   code.
+   * \return The return code, one of `SUCCESS`, `INTERRUPTED`, or `TIMEOUT`.
+   */
+  template<typename Condition, typename DurationT = std::chrono::milliseconds>
+  FutureReturnCode spin_until_complete(
+    const Condition & condition,
+    DurationT timeout = DurationT(-1))
+  {
+    if constexpr (std::is_invocable_v<Condition>) {
+      using RetT = std::invoke_result_t<Condition>;
+      static_assert(
+        std::is_same_v<bool, RetT>,
+        "Conditional callable has to return boolean type");
+      return spin_until_complete_impl(condition, timeout);
+    } else {
+      auto checkFuture = [&condition]() {
+          return condition.wait_for(std::chrono::seconds(0)) ==
+                 std::future_status::ready;
+        };
+      return spin_until_complete_impl(checkFuture, timeout);
+    }
+  }
+
+  /// spin (blocking) for given amount of duration.
+  /**
+   * \param[in] duration gets passed to Executor::spin_node_once,
+   *   spins the executor for given duration.
+   */
+  template<typename DurationT>
+  void
+  spin_for(DurationT duration)
+  {
+    (void)spin_until_complete([]() {return false;}, duration);
+  }
```

Comment:
I'm concerned that the semantics of this method are misleading slightly, and that we need to at least document it more clearly or even consider a name change.

It's inaccurate to say that we will spin for the given duration because we cannot guarantee how long we will spin (i.e. block) for. We can only control how long we wait for work to be available.

Consider this timeline:

- enter `spin_for(1s)`
- wait 0.5 seconds for something to be ready
- execute the callback for that thing for 1 second
- return from `spin_for(1s)` after roughly 1.5 seconds

In my opinion, we should at the very least clarify this in the documentation, and more Ideally pick a name and terminology that reflects what's really happening.

For example, but not necessarily the best answer, you could say `max_wait_time` instead of `duration` or `spin_for_at_least()` instead of `spin_for()`.

Reply:
This is probably the most significant concern I have.

Reply:
It might be sufficient to use phrasing like the `std::this_thread::sleep_for()` function uses:

> Blocks the execution of the current thread for at least the specified sleep_duration.
>
> This function may block for longer than sleep_duration due to scheduling or resource contention delays.

https://en.cppreference.com/w/cpp/thread/sleep_for

Reply:
Sorry for the late reply.

I had the same concern while producing this code. Making it more apparent in a description section would be a must - I've kind of considered that as a user's responsibility to be aware of it, but now I think it might be wrong to do so. 

I was also considering a slight improvement, where we could execute the inner lambda within a separate thread.
But I was thinking about it for some time and I'm not sure whether it's the correct approach - I'm not sure if this would not bring additional overhead. 

I thought that could have an additional parameter or second variant where we directly specify whether the call spawns thread inside or does it in the same thread. Then we could state 2 things, the first approach would not exceed the timeout but lambda could be terminated before it finishes and there will be an additional thread, if you'd rather not then call is without a thread but then you might exceed the timeout. 

I think you've suggested something similar in #1916.

what do you think about it?

Reply:
> I was also considering a slight improvement, where we could execute the inner lambda within a separate thread.
But I was thinking about it for some time and I'm not sure whether it's the correct approach - I'm not sure if this would not bring additional overhead.

👎 The callback should be executed in the thread that called spin, if it's a single threaded executor.

> lambda could be terminated before it finishes

I do not believe there is a way to do that.

> I thought that could have an additional parameter or second variant where we directly specify whether the call spawns thread inside or does it in the same thread. Then we could state 2 things, the first approach would not exceed the timeout but lambda could be terminated before it finishes and there will be an additional thread, if you'd rather not then call is without a thread but then you might exceed the timeout.

I think that's unnecessarily complicated for this function. I'd prefer to let the user do that or maybe have a separate function/class to help them do that.

I think it's just important to ensure the user understands the blocking behavior and that the documentation and/or the name of the function and parameters reflect that.